### PR TITLE
`vault_terraform_cloud_secret_backend` add write-only argument

### PR DESCRIPTION
**Pull request by drewmullen**
### Description

Relates OR Closes https://github.com/hashicorp/terraform-provider-vault/issues/2453

### Checklist
- [x] Added [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md) entry (only for user-facing changes)
- [x] Acceptance tests where run against all supported Vault Versions

### Output from acceptance testing:

```shell
$ TF_ACC=1 go test -v -run TestTerraformCloudSecretBackend -timeout 10m ./...
<redacted>
ok  	github.com/hashicorp/terraform-provider-vault/util/mountutil	(cached) [no tests to run]
=== RUN   TestTerraformCloudSecretBackend
--- PASS: TestTerraformCloudSecretBackend (0.90s)
=== RUN   TestTerraformCloudSecretBackend_remount
--- PASS: TestTerraformCloudSecretBackend_remount (2.07s)
=== RUN   TestTerraformCloudSecretBackend_token_wo
--- PASS: TestTerraformCloudSecretBackend_token_wo (0.87s)
=== RUN   TestTerraformCloudSecretBackendRoleNameFromPath
--- PASS: TestTerraformCloudSecretBackendRoleNameFromPath (0.00s)
=== RUN   TestTerraformCloudSecretBackendRoleBackendFromPath
--- PASS: TestTerraformCloudSecretBackendRoleBackendFromPath (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-vault/vault	4.612s
```

This doesnt seem to do anything but is mentioned in the PR template so I am including. See ACC test runs above:

```shell
$ make testacc TESTARGS='-run=TestTerraformCloudSecretBackend'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test -run=TestTerraformCloudSecretBackend -timeout 30m ./...
?   	github.com/hashicorp/terraform-provider-vault	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/cmd/coverage	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/cmd/generate	[no test files]
ok  	github.com/hashicorp/terraform-provider-vault/codegen	0.192s [no tests to run]
?   	github.com/hashicorp/terraform-provider-vault/helper	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/consts	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/framework/base	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/framework/client	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/framework/errutil	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/framework/model	[no test files]
ok  	github.com/hashicorp/terraform-provider-vault/internal/framework/validators	0.287s [no tests to run]
ok  	github.com/hashicorp/terraform-provider-vault/internal/identity/entity	0.436s [no tests to run]
?   	github.com/hashicorp/terraform-provider-vault/internal/identity/group	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/identity/mfa	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/pki	[no test files]
ok  	github.com/hashicorp/terraform-provider-vault/internal/provider	0.479s [no tests to run]
?   	github.com/hashicorp/terraform-provider-vault/internal/provider/fwprovider	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/providertest	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/rotation	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/sync	[no test files]
ok  	github.com/hashicorp/terraform-provider-vault/internal/vault/secrets/ephemeral	0.576s [no tests to run]
ok  	github.com/hashicorp/terraform-provider-vault/internal/vault/sys	1.005s [no tests to run]
?   	github.com/hashicorp/terraform-provider-vault/schema	[no test files]
ok  	github.com/hashicorp/terraform-provider-vault/testutil	0.538s [no tests to run]
ok  	github.com/hashicorp/terraform-provider-vault/util	0.715s [no tests to run]
ok  	github.com/hashicorp/terraform-provider-vault/util/mountutil	1.041s [no tests to run]
ok  	github.com/hashicorp/terraform-provider-vault/vault	4.874s
...
```

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

### CODE REVIEW

**Comment by Zlaticanin** [comment]
@drewmullen Thank you so much for working on this 🙏 
Could you please resolve the conflicts of the PR?

**Comment by drewmullen** [comment]
rebased:

```shell
TF_ACC=1 go test -v -run TestTerraformCloudSecretBackend -timeout 10m ./...
?   	github.com/hashicorp/terraform-provider-vault	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/cmd/coverage	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/cmd/generate	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-vault/codegen	(cached) [no tests to run]
?   	github.com/hashicorp/terraform-provider-vault/helper	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/consts	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/framework/base	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/framework/client	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/framework/errutil	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/framework/model	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-vault/internal/framework/validators	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-vault/internal/identity/entity	(cached) [no tests to run]
?   	github.com/hashicorp/terraform-provider-vault/internal/identity/group	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/identity/mfa	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/pki	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-vault/internal/provider	(cached) [no tests to run]
?   	github.com/hashicorp/terraform-provider-vault/internal/provider/fwprovider	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/providertest	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/rotation	[no test files]
?   	github.com/hashicorp/terraform-provider-vault/internal/sync	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-vault/internal/vault/secrets/ephemeral	0.967s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-vault/internal/vault/sys	0.672s [no tests to run]
?   	github.com/hashicorp/terraform-provider-vault/schema	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-vault/testutil	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-vault/util	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-vault/util/mountutil	(cached) [no tests to run]
=== RUN   TestTerraformCloudSecretBackend
--- PASS: TestTerraformCloudSecretBackend (0.92s)
=== RUN   TestTerraformCloudSecretBackend_remount
--- PASS: TestTerraformCloudSecretBackend_remount (2.20s)
=== RUN   TestTerraformCloudSecretBackend_token_wo
--- PASS: TestTerraformCloudSecretBackend_token_wo (0.94s)
=== RUN   TestTerraformCloudSecretBackendRoleNameFromPath
--- PASS: TestTerraformCloudSecretBackendRoleNameFromPath (0.00s)
=== RUN   TestTerraformCloudSecretBackendRoleBackendFromPath
--- PASS: TestTerraformCloudSecretBackendRoleBackendFromPath (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-vault/vault	4.863s
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ## 5.2.1 (Aug 19, 2025)
 
+FEATURES:
```
Comment (Zlaticanin) [inline CHANGELOG.md:3]: Can we move this above `## 5.2.1 (Aug 19, 2025)` under `Unreleased`?
And maybe update it to say something along the lines of:

Add support for write-only token argument in `vault_terraform_cloud_secret_backend` resource ([#2483](https://github.com/hashicorp/terraform-provider-vault/pull/2483))

That way we stay consistent with [other feature changelog entries](https://github.com/hashicorp/terraform-provider-vault/blob/47c7a435fddf60bd3fb6f84340a95b9a52ca874d/CHANGELOG.md?plain=1#L21-L25)

File: vault/resource_terraform_cloud_secret_backend.go [hunk 2]
```diff
@@ -177,12 +210,12 @@ func terraformCloudSecretBackendUpdate(ctx context.Context, d *schema.ResourceDa
...
 
-	if d.HasChange("address") || d.HasChange("token") || d.HasChange("base_path") {
+	if d.HasChange(consts.FieldAddress) || d.HasChange(consts.FieldToken) || d.HasChange(consts.FieldBasePath) {
 		log.Printf("[DEBUG] Updating Terraform Cloud configuration at %q", configPath)
 		data := map[string]interface{}{
-			"address":   d.Get("address").(string),
-			"token":     d.Get("token").(string),
-			"base_path": d.Get("base_path").(string),
+			consts.FieldAddress:  d.Get(consts.FieldAddress).(string),
+			consts.FieldToken:    d.Get(consts.FieldToken).(string),
```
Comment (Zlaticanin) [inline vault/resource_terraform_cloud_secret_backend.go]: I see that we always include `d.Get(consts.FieldToken)` when `address` or `base_path` changes. What happens if the resource was created with `token_wo` (so token is empty in state)? Will this overwrite the Vault config with an empty token?

File: vault/resource_terraform_cloud_secret_backend.go [hunk 3]
```diff
@@ -198,6 +231,20 @@ func terraformCloudSecretBackendUpdate(ctx context.Context, d *schema.ResourceDa
 			return diag.FromErr(err)
 		}
 	}
+
+	if d.HasChange(consts.FieldTokenWOVersion) {
+		tokenWO := d.Get(consts.FieldTokenWO).(string)
```
Comment (Zlaticanin) [inline vault/resource_terraform_cloud_secret_backend.go]: Since `tokenWo` is `WriteOnly`, this won't be in the state after the initial apply, correct?
If a user bumps only `token_wo_version` without re-providing `token_wo`, would that write an empty token into Vault?

**Review by Zlaticanin — COMMENTED** [review]
Thank you for working on this!
Since I am not very familiar with this resource, I left some comments/questions to better understand it before approving. 🙏

File: vault/resource_terraform_cloud_secret_backend.go [hunk 4]
```diff
@@ -191,13 +223,45 @@ func terraformCloudSecretBackendUpdate(ctx context.Context, d *schema.ResourceDa
...
-		if err := d.Set("base_path", data["base_path"]); err != nil {
+	}
+
+	if d.HasChange(consts.FieldTokenWOVersion) {
+		log.Printf("[DEBUG] Updating write-only Terraform Cloud token for %q", backend)
+
+		var token string
+		p := cty.GetAttrPath(consts.FieldTokenWO)
+		woVal, _ := d.GetRawConfigAt(p)
+		if !woVal.IsNull() {
```
Comment (Zlaticanin) [inline vault/resource_terraform_cloud_secret_backend.go]: Should we also add a guard here for empty strings? I was thinking something along the lines of:
```
if woVal.IsNull() || woVal.AsString() == "" {
	return diag.Errorf("token_wo must be provided whenever token_wo_version changes")
}
token := woVal.AsString()
```

This way we would resolve the potential panic and prevent empty strings from wiping out the token. Also, it might be a better user experience if they get an error rather than silently breaking the config.

Comment (drewmullen) [reply vault/resource_terraform_cloud_secret_backend.go]: Sure I can make that change! Interestingly the `AsString()` function calls the same `IsNull()`. I'm going to exclude it from the `if`

```go
p := cty.GetAttrPath(consts.FieldTokenWO)
woVal, _ := d.GetRawConfigAt(p)
token := woVal.AsString()
if token == "" {
	return diag.Errorf("token_wo must be provided whenever token_wo_version changes")
}
```

File: vault/resource_terraform_cloud_secret_backend.go [hunk 5]
```diff
@@ -191,13 +223,45 @@ func terraformCloudSecretBackendUpdate(ctx context.Context, d *schema.ResourceDa
...
+
+	if d.HasChange(consts.FieldTokenWOVersion) {
+		log.Printf("[DEBUG] Updating write-only Terraform Cloud token for %q", backend)
+
+		p := cty.GetAttrPath(consts.FieldTokenWO)
+		woVal, _ := d.GetRawConfigAt(p)
+		token := woVal.AsString()
+		if token == "" {
+			return diag.Errorf("token_wo must be provided whenever token_wo_version changes")
+		}
```
Comment (Zlaticanin) [inline vault/resource_terraform_cloud_secret_backend.go]: So, this currently panics because we call `woVal.AsString()` even when the value is null or unknown. Since `AsString()` itself panics on those cases, we crash if `token_wo` isn’t actually provided in the plan. I was able to test this by writing a case where I bumped `token_wo_version` without including `token_wo`, and it panicked right away.

What do you think about first checking that `token_wo_version` is explicitly set, and then validate that `token_wo` is both known and non-null before calling `AsString()`? That way, we return a clean error message if it’s missing or empty, instead of panicking. I tested this version too, and it behaved as expected, meaning no panics, errors when `token_wo` was omitted or empty, and also successful when a valid value was provided.

```
if d.HasChange(consts.FieldTokenWOVersion) {
		versionVal := d.GetRawConfig().GetAttr(consts.FieldTokenWOVersion)
		if versionVal.IsKnown() && !versionVal.IsNull() {
			woVal := d.GetRawConfig().GetAttr(consts.FieldTokenWO)
			if !woVal.IsKnown() || woVal.IsNull() || strings.TrimSpace(woVal.AsString()) == "" {
				return diag.Errorf("token_wo must be provided whenever token_wo_version changes")
			}
			token := woVal.AsString()
			if _, err := client.Logical().WriteWithContext(ctx, configPath, map[string]interface{}{
				consts.FieldToken: token,
			}); err != nil {
				return diag.Errorf("Error configuring Terraform Cloud configuration for %q: %s", backend, err)
			}
		}
	}
```

File: vault/resource_terraform_cloud_secret_backend_test.go [hunk 6]
```diff
@@ -98,6 +98,36 @@ func TestTerraformCloudSecretBackend_remount(t *testing.T) {
 	})
 }
 
+func TestTerraformCloudSecretBackend_token_wo(t *testing.T) {
```
Comment (Zlaticanin) [inline vault/resource_terraform_cloud_secret_backend_test.go]: Right now this test only covers the initial create path and the happy path when we update `token_wo` fields. We’re not testing edge cases like setting `token_wo_version` without `token_wo` which should result in an error, as well as providing `token_wo = ""` which should also result in an error. We should probably add those cases to make sure that all cases are validated.

File: vault/resource_terraform_cloud_secret_backend.go (hunk 4 above)
Comment (Zlaticanin) [reply vault/resource_terraform_cloud_secret_backend.go]: Closing this convo as I made another comment with the current code.

File: CHANGELOG.md (hunk 1 above)
Comment (Zlaticanin) [reply CHANGELOG.md:3]: Small nit: Would you mind re-wording the changelog entry so it stays consistent with others? Thank you 🙏

File: vault/resource_terraform_cloud_secret_backend_test.go [hunk 7]
```diff
@@ -98,6 +99,50 @@ func TestTerraformCloudSecretBackend_remount(t *testing.T) {
...
+			},
+			// This test case is to test the validation logic of token_wo and token_wo_version
+			// We expect an error because token_wo is missing, it requires specific error because of the
+			// RequiredWith setting in the schema
+			{
+				Config:      testTerraformCloudSecretBackend_tokenWoNoVersion(backend, 3),
+				ExpectError: regexp.MustCompile(`all of.*token_wo,token_wo_version.*must be specified`),
+			},
+			// This test case is to test the validation logic of token_wo and token_wo_version when token_wo is ""
+			// where we expect an error
```
Comment (Zlaticanin) [inline vault/resource_terraform_cloud_secret_backend_test.go]:
```suggestion
```

File: vault/resource_terraform_cloud_secret_backend_test.go [hunk 8]
```diff
@@ -98,6 +99,50 @@ func TestTerraformCloudSecretBackend_remount(t *testing.T) {
...
+			{
+				Config: testTerraformCloudSecretBackend_tokenWoUpdatedConfig(backend, token, 2),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "backend", backend),
+					resource.TestCheckResourceAttr(resourceName, "token_wo_version", "2"),
+				),
+			},
+			// This test case is to test the validation logic of token_wo and token_wo_version
+			// We expect an error because token_wo is missing, it requires specific error because of the
+			// RequiredWith setting in the schema
```
Comment (Zlaticanin) [inline vault/resource_terraform_cloud_secret_backend_test.go]:
```suggestion
```

**Comment by drewmullen** [comment]
closing for: https://github.com/hashicorp/terraform-provider-vault/pull/2603
